File: src/banks/Bank.ts
import { format } from "date-fns";
import { randomUUID } from "node:crypto";
import { mkdtemp, readFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { BrowserContext, chromium, LaunchOptions, Page } from "playwright";
import { z } from "zod";
import env from "../utils/env";
import logger from "../utils/logger";
import { sendNotification } from "../utils/pushover";
import { uploadFile } from "../utils/s3";
import { Account, BankName, bankNames } from "./types";

export class Bank {
  private readonly bank: BankName;
  private context: BrowserContext | null = null;
  private page: Page | null = null;
  protected date = new Date();
  private accounts: z.infer<typeof Account>[] = [];
  private userDataDir: string | null = null;

  constructor(bank: BankName) {
    this.bank = bank;
  }

  protected async launchBrowser() {
    this.userDataDir = await mkdtemp(join(tmpdir(), `user-data-`));
    logger.debug(`Created temporary user data directory: ${this.userDataDir}`);

    logger.debug("Launching browser");
    const options: LaunchOptions = {
      headless: false,
      args: [
        "--no-sandbox",
        "--disable-setuid-sandbox",
        "--disable-blink-features=AutomationControlled",
      ],
    };
    if (env.HTTP_PROXY) {
      options.proxy = {
        server: env.HTTP_PROXY,
      };
    }
    this.context = await chromium.launchPersistentContext(
      this.userDataDir,
      options,
    );
    await this.startTracing();
    logger.debug("Creating new page");
    this.page = await this.context.newPage();
  }

  protected async closeBrowser(tracingFilePath?: string) {
    await this.stopTracing(tracingFilePath);
    logger.debug("Closing browser");
    await this.page?.context().browser()?.close();
    this.page = null;
  }

  protected async startTracing() {
    logger.debug("Starting tracing");
    await this.context?.tracing.start({ screenshots: true, snapshots: true });
  }

  protected async stopTracing(filePath?: string) {
    logger.debug("Stopping tracing");
    await this.context?.tracing.stop({ path: filePath });
  }

  protected async handleError(error: Error) {
    logger.error(error);
    const errorMessage = error.stack?.split("\n")[0];
    const getTraceFilePath = (fileName: string) => `traces/${fileName}`;
    const getTraceFileName = () =>
      `${format(this.date, "yyyy-MM-dd")}-${this.bank}-${randomUUID()}.zip`;
    const traceFileName = getTraceFileName();
    const traceFilePath = getTraceFilePath(traceFileName);
    await this.closeBrowser(traceFilePath);
    logger.info(`Saved trace to ${traceFilePath}`);
    const traceFile = await readFile(traceFilePath);
    await uploadFile(
      env.AWS_S3_TRACES_BUCKET_NAME,
      traceFileName,
      "application/zip",
      traceFile,
    );
    await sendNotification(errorMessage || null, {
      title: `Error Logging Into ${bankNames[this.bank]}`,
      url: "https://console.aws.amazon.com/cloudwatch/home#logsV2:log-groups",
      url_title: "Open AWS Console",
      priority: -1,
    });
  }

  protected async getCookies() {
    if (!this.page) {
      throw new Error("Page is not initialized");
    }
    return this.page.context().cookies();
  }

  protected async getCookiesAsString() {
    const cookies = await this.getCookies();
    return cookies.map((cookie) => `${cookie.name}=${cookie.value}`).join("; ");
  }

  protected async getCookie(name: string) {
    const cookies = await this.getCookies();
    const cookie = cookies.find((cookie) => cookie.name === name)?.value;
    if (!cookie) {
      throw new Error(`Cookie "${name}" not found`);
    }
    return cookie;
  }

  protected async getPage() {
    if (!this.page) {
      throw new Error("Page is not initialized");
    }
    return this.page;
  }

  public getAccounts() {
    return this.accounts;
  }

  protected setAccounts(accounts: z.infer<typeof Account>[]) {
    this.accounts = accounts;
  }
}

File: src/banks/Bank.py
import datetime
import uuid
import asyncio
from playwright.async_api import async_playwright
from dateutil import parser
from dateutil.relativedelta import relativedelta
from typing import List, Dict, Any
import logging
from utils.pushover import send_notification
from utils.s3 import upload_file
from banks.types import Account, bankName, BankName

class Bank:
    def __init__(self, bank: BankName):
        self.bank = bank
        self.context = None
        self.page = None
        self.date = datetime.datetime.now()
        self.accounts: List[Account] = []

    async def launch_browser(self):
        logging.debug("Launching browser")
        options = {
            "headless": False,
            "args": [
                "--no-sandbox",
                "--disable-setuid-sandbox",
                "--disable-blink-features=AutomationControlled",
            ],
        }
        async with async_playwright() as p:
            browser = await p.chromium.launch(**options)
            logging.debug("Creating new context")
            self.context = await browser.new_context()
            await self.start_tracing()
            logging.debug("Creating new page")
            self.page = await self.context.new_page()

    async def close_browser(self, tracing_file_path: str = None):
        await self.stop_tracing(tracing_file_path)
        logging.debug("Closing browser")
        await self.page.context.browser.close()
        self.page = None

    async def start_tracing(self):
        logging.debug("Starting tracing")
        await self.context.tracing.start(screenshots=True, snapshots=True)

    async def stop_tracing(self, file_path: str = None):
        logging.debug("Stopping tracing")
        await self.context.tracing.stop(path=file_path)

    async def handle_error(self, error: Exception):
        logging.error(error)
        trace_file_name = f"{self.date.strftime('%Y-%m-%d')}-{self.bank}-{uuid.uuid4()}.zip"
        trace_file_path = f"traces/{trace_file_name}"
        await self.close_browser(trace_file_path)
        logging.info(f"Saved trace to {trace_file_path}")
        with open(trace_file_path, "rb") as trace_file:
            await upload_file(trace_file_name, "application/zip", trace_file.read())
        await send_notification(
            f"Error fetching accounts from {bankName[self.bank]}.",
            {
                "title": "Error Fetching Accounts",
                "url": "https://console.aws.amazon.com/cloudwatch/home#logsV2:log-groups",
                "url_title": "Open AWS Console",
            },
        )

    async def get_cookies(self):
        if not self.page:
            raise Exception("Page is not initialized")
        return await self.page.context.cookies()

    async def get_cookies_as_string(self):
        cookies = await self.get_cookies()
        return "; ".join([f"{cookie['name']}={cookie['value']}" for cookie in cookies])

    async def get_cookie(self, name: str):
        cookies = await self.get_cookies()
        cookie = next((cookie['value'] for cookie in cookies if cookie['name'] == name), None)
        if not cookie:
            raise Exception(f'Cookie "{name}" not found')
        return cookie

    async def get_page(self):
        if not self.page:
            raise Exception("Page is not initialized")
        return self.page

    def get_accounts(self):
        return self.accounts

    def set_accounts(self, accounts: List[Account]):
        self.accounts = accounts
